tcpip/config/doc/ip_configurations.md
# Internet Protocol (IP) Configurations

**IPv4 Specific User Configurations**

	
- **Enable Console Commands:**
    - Enable the IPv4 TCP/IP Console Commands
    - Default value: false
    - Symbol: TCPIP_IPV4_COMMANDS_ENABLE, TCPIP_IPV4_COMMANDS



- **Number of ARP Requests IPv4 can queue:**
    - The number of entries that IPv4 can queue up for ARP resolution.
    - Usually it should be <= the number of total ARP cache entries for all interfaces
    - Default value: 10
    - Symbol TCPIP_IPV4_ARP_QUEUE_SLOTS, TCPIP_IPV4_ARP_SLOTS

- **Enable IPv4 Fragmentation Support:**
    - If 1, enable IPv4 fragmentation processing, RX + TX
    - If 0 (default), IPv4 fragmentation not supported
    - Default value: false
    - Symbol TCPIP_IPV4_FRAGMENTATION


- **Initial fragment timeout (in sec):**
    - Initial fragment timeout
    - Timer Lower Bound (RFC 791)
    - The RFC states this value to be 15 seconds
    - Value is in seconds
    - Default value: 15
    - Symbol TCPIP_IPV4_FRAGMENT_TIMEOUT

- **Maximum number of fragmented streams:**
    - Upper limit for the number of fragmented streams that could be handled simultaneously
    - Default value: 3
    - Symbol TCPIP_IPV4_FRAGMENT_MAX_STREAMS

- **Maximum number of fragments:**
    - Upper limit for the number of fragments in a fragmented stream
    - This also imposes a limit on how large a packet could be depending on the MTU and fragmentations that occur
    - If more fragments received for a certain stream, the stream will be discarded
    - Default value: 4
    - Symbol TCPIP_IPV4_FRAGMENT_MAX_NUMBER

- **Enable External Packet Processing:**
    - This setting enables/disables the processing of the RX packets by an external handler
    - The user of the IPv4 can register an external function to process the incoming packets
    - If true, the functionality is built in and could be used by the application
    - If false, the functionality does not exist and the generated code is slightly smaller 
    - Default value: false
    - Symbol TCPIP_IPV4_EXTERN_PACKET_PROCESS


- **IPv4 Fragmentation Task Rate (in msec):**
    - IPv4 task processing rate, in milliseconds.
    - The IPv4 module will process a timer event with this rate for maintaining its own queues, processing timeouts, etc.
    - The lower the rate (higher the frequency) the higher the module priority and higher module performance can be obtained
    - The default value is around 30 milliseconds.
    - The value cannot be lower than the TCPIP_STACK_TICK_RATE.
    - Value is in milliseconds
    - Default value: 37
    - Symbol TCPIP_IPV4_TASK_TICK_RATE

- **Enable IPv4 Forwarding Support:**
    - Builds the IPv4 module with forwarding support
    - Default value: false
    - Symbol TCPIP_IPV4_FORWARD, TCPIP_IPV4_FORWARDING_ENABLE

- **Support Forwarding Table with ASCII strings:**
    - The initialization forwarding table is in ASCII format, using strings
    - Default value: true
    - Symbol TCPIP_IPV4_FORWARD_ASCII_ENABLE, TCPIP_IPV4_FORWARDING_TABLE_ASCII

- **Support Dynamic Forwarding Table Manipulation:**
    - Enables run time API for forwarding tables manipulation
    - Certain API functions (TCPIP_IPV4_FwdTableAddAscii, TCPIP_IPV4_FwdTableRemoveAscii, ...) exists only if they are enabled
    - Default value: true
    - Symbol TCPIP_IPV4_FORWARD_DYN_ENABLE, TCPIP_IPV4_FORWARDING_DYNAMIC_API


- **Enable Broadcast Traffic Forwarding:**
    - Enable the forwarding of the broadcast packets
    - The default should be disabled
    - Default value: false
    - Symbol TCPIP_IPV4_FORWARD_BCAST_ENABLE


- **Enable Forwarding Statistics:**
    - Enable the IPv4 module to keep statistics counts
    - Default value: false
    - Symbol TCPIP_IPV4_FORWARD_STATS_ENABLE, TCPIP_IPV4_FORWARDING_STATS

- **Forwarding TX Queue Slots:**
    - The maximum number of broadcast and multicast packets that can be queued waiting for the output interface to transmit them.
    - This applies only for packets that have to be both forwarded and processed internally which is only broadcast or multicast
    - Adjust depending on your traffic
    - Note that if this limit is exceeded, the packets will not be forwarded but still processed internally
    - If 0, packets will not be forwarded, just processed internally.
    - Default value: 0
    - Symbol TCPIP_IPV4_FWD_QUEUE_SLOTS

- **Forwarding Table Maximum Size:**
    - The maximum number of entries in the forwarding table for each interface
    - Default value: 10
    - Symbol TCPIP_IPV4_FORWARD_TABLE_MAX_SIZE, TCPIP_IPV4_FORWARDING_TABLE_MAX_SIZE

- **Use ASCII Init Forwarding Table:**
    - Initialization table built with ASCII strings
    - Default value: true
    - Symbol TCPIP_IPV4_FORWARD_ASCII_SELECT

- **Number of Entries in the Init Forwarding table:**
    - Number of entries in the routing table as part of the initialization
    - Default value: none
    - Symbol TCPIP_IPV4_FORWARD_TABLE_CNT, TCPIP_IPV4_FORWARDING_TABLE_ENTRIES


- **Network Destination Address:**
    - Network destination address.
    - This address is used for matching the IPv4 packet destination address. For ex: "192.168.2.0"
    - Default value: "0.0.0.0" 
    - Symbol TCPIP_IPV4_FWD_ENTRY_NET_ADD_IDX

- **Network Mask:**
    - Associated mask for this destination address. For ex: "255.255.255.0"
    - The value needs to be a valid network mask, with contiguous ones in its binary representation followed by contiguous zeros
    - Default value: "0.0.0.0" 
    - Symbol TCPIP_IPV4_FWD_ENTRY_NET_MASK_IDX


- **Gateway Destination Address:**
    - Gateway destination address. For ex: "192.168.2.1"
    - Used as packets destination */
    - Default value: "0.0.0.0" 
    - Symbol TCPIP_IPV4_FWD_ENTRY_GW_ADD_IDX

- **Source Interface Name:**
    - The name of the input interface that this entry applies to.
    - The forwarding tables are built per interface. For ex: "eth0", "wlan0", etc.
    - Default value: ""
    - Symbol TCPIP_IPV4_FWD_ENTRY_SRC_IF_NAME_IDX

- **Destination Interface Name:**
    - The name of the interface to go out on if this entry is selected.  For ex: "eth0", "wlan0", etc.
    - Default value: ""
    - Symbol TCPIP_IPV4_FWD_ENTRY_DEST_IF_NAME_IDX

- **Entry Metric Value:**
    - The path efficiency. The smaller the value, the higher priority for the path.
    - Note that this value is useful only when there are multiple entries in the forwarding table that have the same number of mask bits set
    - Default value: 10
    - Symbol TCPIP_IPV4_FWD_ENTRY_METRIC_IDX



tcpip/config/doc/arp_configurations.md
# Address Resolution Protocol (ARP) Configurations 

**ARP Specific User Configurations**

- **Number of Cache Entries:**
    - Default number of entries per interface.
    - Default value 5
    - Symbol TCPIP_ARP_CACHE_ENTRIES


- **Enable ARP Commands:**
    - Enable the ARP TCP/IP Console Commands
    - Default value: false
    - Symbol: TCPIP_ARP_COMMANDS_ENABLE


- **Delete Old Cache:**
    - On initialization, delete the old cache if still in place
    - Else do not re-initialize
    - Default value: 1
    - Symbol TCPIP_ARP_CACHE_DELETE_OLD

- **Time-Out for removal of Solved Entry (in sec):**
    - Timeout for a solved entry in the cache.
    - The entry will be removed if the tmo elapsed and the entry has not been referenced again
    - Value is in seconds
    - Default value: (20 * 60)
    - Symbol TCPIP_ARP_CACHE_SOLVED_ENTRY_TMO


- **Time-Out to solve Pending Cache Entry (in sec):**
    - Timeout for a cache entry pending to be solved
    - The entry will be removed if the tmo elapsed and the entry has not been solved.
    - A solved entry moves to the solved entries timeout.
    - Value is in seconds
    - Default value: (1 * 60)
    - Symbol TCPIP_ARP_CACHE_PENDING_ENTRY_TMO


- **Time-Out for Resending an ARP Request for Pending Entry (in sec):**
    - Timeout for resending an ARP request for a pending entry
    - In order to prevent the ARP flooding the standard recommends it to be greater than 1 sec.
    - It should be less than TCPIP_ARP_CACHE_PENDING_ENTRY_TMO  
    - Value is in seconds
    - Default value: 2
    - Symbol TCPIP_ARP_CACHE_PENDING_RETRY_TMO

    
- **Number of Retries for Resolving an Entry:**
    - Number of ARP requests generated for resolving an entry.
    - Default value: 3
    - Symbol TCPIP_ARP_CACHE_ENTRY_RETRIES


- **Number of Retries for Gratuitous ARP Request:**
    - Number of ARP requests generated when sending a gratuitous ARP probe.
    - Default value: 1
    - Symbol TCPIP_ARP_GRATUITOUS_PROBE_COUNT


- **Purge Threshold Percentage:**
    - Default purge threshold, percentage
    - Once the number of resolved entries in the cache gets beyond the threshold some resolved entries will be purged.
    - Default value: 75
    - Symbol TCPIP_ARP_CACHE_PURGE_THRESHOLD

- **Number of Entries to Delete Once the Threshold is Reached:**
    - The number of entries to delete, once the threshold is reached.
    - Default value: 3
    - Symbol TCPIP_ARP_CACHE_PURGE_QUANTA


- **Maximum Percentage of Permanent Entries Allowed in Cache:**
    - Maximum percentage of permanent entries in the cache.
    - Note that since permanent entries cannot be removed they tend to degrade the efficiency of the cache look up.  
    - Default value: 50
    - Symbol TCPIP_ARP_CACHE_PERMANENT_QUOTA


- **Alias Interfaces Share One Cache:**
    - ARP caches configuration for aliases.
    - If this symbol is true, then the alias interfaces share the cache with the primary interface.
    - Otherwise each interface uses its own cache.
    - Default value should be 1 if the alias interfaces are in the same network with the primary interface
    - Symbol TCPIP_ARP_PRIMARY_CACHE_ONLY


- **ARP Task Rate (in msec):**
    - ARP task processing rate, in milliseconds.
    - The ARP module will process a timer event with this rate  for maintaining its own queues, processing timeouts, etc.
    - Choose it so that the other ARP_CACHE_xxx_TMO are multiple of this
    - The lower the rate (higher the frequency) the higher the module priority and higher module performance can be obtained
    - The value cannot be lower than the TCPIP_STACK_TICK_RATE.
    - Default value: 2000
    - Symbol TCPIP_ARP_TASK_PROCESS_RATE























apps/iperf_demo/firmware/src/config/sam_e70_xult/configuration.h
/*******************************************************************************
  System Configuration Header

  File Name:
    configuration.h

  Summary:
    Build-time configuration header for the system defined by this project.

  Description:
    An MPLAB Project may have multiple configurations.  This file defines the
    build-time options for a single configuration.

  Remarks:
    This configuration header must not define any prototypes or data
    definitions (or include any files that do).  It only provides macro
    definitions for build-time configuration options

*******************************************************************************/

// DOM-IGNORE-BEGIN
/*******************************************************************************
* Copyright (C) 2018 Microchip Technology Inc. and its subsidiaries.
*
* Subject to your compliance with these terms, you may use Microchip software
* and any derivatives exclusively with Microchip products. It is your
* responsibility to comply with third party license terms applicable to your
* use of third party software (including open source software) that may
* accompany Microchip software.
*
* THIS SOFTWARE IS SUPPLIED BY MICROCHIP "AS IS". NO WARRANTIES, WHETHER
* EXPRESS, IMPLIED OR STATUTORY, APPLY TO THIS SOFTWARE, INCLUDING ANY IMPLIED
* WARRANTIES OF NON-INFRINGEMENT, MERCHANTABILITY, AND FITNESS FOR A
* PARTICULAR PURPOSE.
*
* IN NO EVENT WILL MICROCHIP BE LIABLE FOR ANY INDIRECT, SPECIAL, PUNITIVE,
* INCIDENTAL OR CONSEQUENTIAL LOSS, DAMAGE, COST OR EXPENSE OF ANY KIND
* WHATSOEVER RELATED TO THE SOFTWARE, HOWEVER CAUSED, EVEN IF MICROCHIP HAS
* BEEN ADVISED OF THE POSSIBILITY OR THE DAMAGES ARE FORESEEABLE. TO THE
* FULLEST EXTENT ALLOWED BY LAW, MICROCHIP'S TOTAL LIABILITY ON ALL CLAIMS IN
* ANY WAY RELATED TO THIS SOFTWARE WILL NOT EXCEED THE AMOUNT OF FEES, IF ANY,
* THAT YOU HAVE PAID DIRECTLY TO MICROCHIP FOR THIS SOFTWARE.
*******************************************************************************/
// DOM-IGNORE-END

#ifndef CONFIGURATION_H
#define CONFIGURATION_H

// *****************************************************************************
// *****************************************************************************
// Section: Included Files
// *****************************************************************************
// *****************************************************************************
/*  This section Includes other configuration headers necessary to completely
    define this configuration.
*/

#include "user.h"
#include "toolchain_specifics.h"

// DOM-IGNORE-BEGIN
#ifdef __cplusplus  // Provide C++ Compatibility

extern "C" {

#endif
// DOM-IGNORE-END

// *****************************************************************************
// *****************************************************************************
// Section: System Configuration
// *****************************************************************************
// *****************************************************************************



// *****************************************************************************
// *****************************************************************************
// Section: System Service Configuration
// *****************************************************************************
// *****************************************************************************
/* TIME System Service Configuration Options */
#define SYS_TIME_INDEX_0                     0
#define SYS_TIME_MAX_TIMERS                  5
#define SYS_TIME_HW_COUNTER_WIDTH            16
#define SYS_TIME_HW_COUNTER_PERIOD           65535U
#define SYS_TIME_HW_COUNTER_HALF_PERIOD	     (SYS_TIME_HW_COUNTER_PERIOD>>1)
#define SYS_TIME_CPU_CLOCK_FREQUENCY         300000000
#define SYS_TIME_COMPARE_UPDATE_EXECUTION_CYCLES      (900)

#define SYS_CONSOLE_INDEX_0                       0





#define SYS_CMD_ENABLE
#define SYS_CMD_DEVICE_MAX_INSTANCES       SYS_CONSOLE_DEVICE_MAX_INSTANCES
#define SYS_CMD_PRINT_BUFFER_SIZE          1024
#define SYS_CMD_BUFFER_DMA_READY



#define SYS_DEBUG_ENABLE
#define SYS_DEBUG_GLOBAL_ERROR_LEVEL       SYS_ERROR_DEBUG
#define SYS_DEBUG_BUFFER_DMA_READY
#define SYS_DEBUG_USE_CONSOLE


#define SYS_CONSOLE_DEVICE_MAX_INSTANCES   			1
#define SYS_CONSOLE_UART_MAX_INSTANCES 	   			1
#define SYS_CONSOLE_USB_CDC_MAX_INSTANCES 	   		0
#define SYS_CONSOLE_PRINT_BUFFER_SIZE        		200




// *****************************************************************************
// *****************************************************************************
// Section: Driver Configuration
// *****************************************************************************
// *****************************************************************************


/*** MIIM Driver Configuration ***/
#define DRV_MIIM_ETH_MODULE_ID              GMAC_BASE_ADDRESS
#define DRV_MIIM_INSTANCES_NUMBER           1
#define DRV_MIIM_INSTANCE_OPERATIONS        4
#define DRV_MIIM_INSTANCE_CLIENTS           2
#define DRV_MIIM_CLIENT_OP_PROTECTION   false
#define DRV_MIIM_COMMANDS   false
#define DRV_MIIM_DRIVER_OBJECT              DRV_MIIM_OBJECT_BASE_Default
#define DRV_MIIM_DRIVER_INDEX               DRV_MIIM_INDEX_0              





// *****************************************************************************
// *****************************************************************************
// Section: Middleware & Other Library Configuration
// *****************************************************************************
// *****************************************************************************

/*** ICMPv4 Server Configuration ***/
#define TCPIP_STACK_USE_ICMP_SERVER
#define TCPIP_ICMP_ECHO_ALLOW_BROADCASTS    false



/*** NBNS Configuration ***/
#define TCPIP_STACK_USE_NBNS
#define TCPIP_NBNS_TASK_TICK_RATE   110



/*** TCP Configuration ***/
#define TCPIP_TCP_MAX_SEG_SIZE_TX		        	1460
#define TCPIP_TCP_SOCKET_DEFAULT_TX_SIZE			512
#define TCPIP_TCP_SOCKET_DEFAULT_RX_SIZE			512
#define TCPIP_TCP_DYNAMIC_OPTIONS             			true
#define TCPIP_TCP_START_TIMEOUT_VAL		        	1000
#define TCPIP_TCP_DELAYED_ACK_TIMEOUT		    		100
#define TCPIP_TCP_FIN_WAIT_2_TIMEOUT		    		5000
#define TCPIP_TCP_KEEP_ALIVE_TIMEOUT		    		10000
#define TCPIP_TCP_CLOSE_WAIT_TIMEOUT		    		0
#define TCPIP_TCP_MAX_RETRIES		            		5
#define TCPIP_TCP_MAX_UNACKED_KEEP_ALIVES			6
#define TCPIP_TCP_MAX_SYN_RETRIES		        	3
#define TCPIP_TCP_AUTO_TRANSMIT_TIMEOUT_VAL			40
#define TCPIP_TCP_WINDOW_UPDATE_TIMEOUT_VAL			200
#define TCPIP_TCP_MAX_SOCKETS		                10
#define TCPIP_TCP_TASK_TICK_RATE		        	5
#define TCPIP_TCP_MSL_TIMEOUT		        	    0
#define TCPIP_TCP_QUIET_TIME		        	    0
#define TCPIP_TCP_COMMANDS   false
#define TCPIP_TCP_EXTERN_PACKET_PROCESS   false



/*** DHCP Configuration ***/
#define TCPIP_STACK_USE_DHCP_CLIENT
#define TCPIP_DHCP_TIMEOUT                          10
#define TCPIP_DHCP_TASK_TICK_RATE                   5
#define TCPIP_DHCP_HOST_NAME_SIZE                   20
#define TCPIP_DHCP_CLIENT_CONNECT_PORT              68
#define TCPIP_DHCP_SERVER_LISTEN_PORT               67
#define TCPIP_DHCP_CLIENT_ENABLED                   true
#define TCPIP_DHCP_USE_OPTION_TIME_SERVER           0
#define TCPIP_DHCP_TIME_SERVER_ADDRESSES            0
#define TCPIP_DHCP_USE_OPTION_NTP_SERVER            0
#define TCPIP_DHCP_NTP_SERVER_ADDRESSES             0



/*** ARP Configuration ***/
#define TCPIP_ARP_CACHE_ENTRIES                 		5
#define TCPIP_ARP_CACHE_DELETE_OLD		        	true
#define TCPIP_ARP_CACHE_SOLVED_ENTRY_TMO			1200
#define TCPIP_ARP_CACHE_PENDING_ENTRY_TMO			60
#define TCPIP_ARP_CACHE_PENDING_RETRY_TMO			2
#define TCPIP_ARP_CACHE_PERMANENT_QUOTA		    		50
#define TCPIP_ARP_CACHE_PURGE_THRESHOLD		    		75
#define TCPIP_ARP_CACHE_PURGE_QUANTA		    		1
#define TCPIP_ARP_CACHE_ENTRY_RETRIES		    		3
#define TCPIP_ARP_GRATUITOUS_PROBE_COUNT			1
#define TCPIP_ARP_TASK_PROCESS_RATE		        	2
#define TCPIP_ARP_PRIMARY_CACHE_ONLY		        	true



/*** tcpip_cmd Configuration ***/
#define TCPIP_STACK_COMMAND_ENABLE
#define TCPIP_STACK_COMMANDS_ICMP_ECHO_REQUESTS         4
#define TCPIP_STACK_COMMANDS_ICMP_ECHO_REQUEST_DELAY    1000
#define TCPIP_STACK_COMMANDS_ICMP_ECHO_TIMEOUT          5000
#define TCPIP_STACK_COMMANDS_WIFI_ENABLE             	false
#define TCPIP_STACK_COMMANDS_ICMP_ECHO_REQUEST_BUFF_SIZE    2000
#define TCPIP_STACK_COMMANDS_ICMP_ECHO_REQUEST_DATA_SIZE    100



/* Network Configuration Index 0 */
#define TCPIP_NETWORK_DEFAULT_INTERFACE_NAME_IDX0	"GMAC"
#define TCPIP_IF_GMAC

#define TCPIP_NETWORK_DEFAULT_HOST_NAME_IDX0				"MCHPBOARD_C"
#define TCPIP_NETWORK_DEFAULT_MAC_ADDR_IDX0				"00:04:25:1C:A0:02"

#define TCPIP_NETWORK_DEFAULT_IP_ADDRESS_IDX0			"192.168.100.11"
#define TCPIP_NETWORK_DEFAULT_IP_MASK_IDX0			"255.255.255.0"
#define TCPIP_NETWORK_DEFAULT_GATEWAY_IDX0			"192.168.100.1"
#define TCPIP_NETWORK_DEFAULT_DNS_IDX0				"192.168.100.1"
#define TCPIP_NETWORK_DEFAULT_SECOND_DNS_IDX0			"0.0.0.0"
#define TCPIP_NETWORK_DEFAULT_POWER_MODE_IDX0			"full"
#define TCPIP_NETWORK_DEFAULT_INTERFACE_FLAGS_IDX0			\
													TCPIP_NETWORK_CONFIG_DHCP_CLIENT_ON |\
													TCPIP_NETWORK_CONFIG_DNS_CLIENT_ON |\
													TCPIP_NETWORK_CONFIG_IP_STATIC
													
#define TCPIP_NETWORK_DEFAULT_MAC_DRIVER_IDX0			DRV_GMAC_Object



/*** iperf Configuration ***/
#define TCPIP_STACK_USE_IPERF
#define TCPIP_IPERF_TX_BUFFER_SIZE		8192
#define TCPIP_IPERF_RX_BUFFER_SIZE  	8192
#define TCPIP_IPERF_TX_WAIT_TMO     	100
#define TCPIP_IPERF_TX_QUEUE_LIMIT  	2
#define TCPIP_IPERF_TIMING_ERROR_MARGIN 0
#define TCPIP_IPERF_MAX_INSTANCES       1
#define TCPIP_IPERF_TX_BW_LIMIT  		1



/*** IPv4 Configuration ***/
#define TCPIP_IPV4_EXTERN_PACKET_PROCESS   false



/*** TCPIP Heap Configuration ***/

#define TCPIP_STACK_USE_INTERNAL_HEAP
#define TCPIP_STACK_DRAM_SIZE                       69250
#define TCPIP_STACK_DRAM_RUN_LIMIT                  2048

#define TCPIP_STACK_MALLOC_FUNC                     malloc

#define TCPIP_STACK_CALLOC_FUNC                     calloc

#define TCPIP_STACK_FREE_FUNC                       free



#define TCPIP_STACK_HEAP_USE_FLAGS                   TCPIP_STACK_HEAP_FLAG_ALLOC_UNCACHED

#define TCPIP_STACK_HEAP_USAGE_CONFIG                TCPIP_STACK_HEAP_USE_DEFAULT

#define TCPIP_STACK_SUPPORTED_HEAPS                  1




// *****************************************************************************
// *****************************************************************************
// Section: TCPIP Stack Configuration
// *****************************************************************************
// *****************************************************************************

#define TCPIP_STACK_USE_IPV4
#define TCPIP_STACK_USE_TCP
#define TCPIP_STACK_USE_UDP

#define TCPIP_STACK_TICK_RATE		        		5
#define TCPIP_STACK_SECURE_PORT_ENTRIES             10

#define TCPIP_STACK_ALIAS_INTERFACE_SUPPORT   false

#define TCPIP_PACKET_LOG_ENABLE     0

/* TCP/IP stack event notification */
#define TCPIP_STACK_USE_EVENT_NOTIFICATION
#define TCPIP_STACK_USER_NOTIFICATION   false
#define TCPIP_STACK_DOWN_OPERATION   true
#define TCPIP_STACK_IF_UP_DOWN_OPERATION   true
#define TCPIP_STACK_MAC_DOWN_OPERATION  true
#define TCPIP_STACK_INTERFACE_CHANGE_SIGNALING   false
#define TCPIP_STACK_CONFIGURATION_SAVE_RESTORE   true
#define TCPIP_STACK_EXTERN_PACKET_PROCESS   false






/*** TCPIP MAC Configuration ***/
#define TCPIP_GMAC_TX_DESCRIPTORS_COUNT_DUMMY				1
#define TCPIP_GMAC_RX_DESCRIPTORS_COUNT_DUMMY				1
#define TCPIP_GMAC_RX_BUFF_SIZE_DUMMY				    	64
#define TCPIP_GMAC_TX_BUFF_SIZE_DUMMY				    	64

/*** QUEUE 0 Configuration ***/
#define TCPIP_GMAC_TX_DESCRIPTORS_COUNT_QUE0				10
#define TCPIP_GMAC_RX_DESCRIPTORS_COUNT_QUE0				10
#define TCPIP_GMAC_RX_BUFF_SIZE_QUE0				    	1536
#define TCPIP_GMAC_TX_BUFF_SIZE_QUE0				    	1536
#define TCPIP_GMAC_RX_BUFF_COUNT_QUE0				   		12
#define TCPIP_GMAC_RX_BUFF_COUNT_THRESHOLD_QUE0				1
#define TCPIP_GMAC_RX_BUFF_ALLOC_COUNT_QUE0					1

/*** QUEUE 1 Disabled; Dummy Configuration ***/
#define TCPIP_GMAC_TX_DESCRIPTORS_COUNT_QUE1				TCPIP_GMAC_TX_DESCRIPTORS_COUNT_DUMMY
#define TCPIP_GMAC_RX_DESCRIPTORS_COUNT_QUE1				TCPIP_GMAC_RX_DESCRIPTORS_COUNT_DUMMY
#define TCPIP_GMAC_RX_BUFF_SIZE_QUE1				    	TCPIP_GMAC_RX_BUFF_SIZE_DUMMY
#define TCPIP_GMAC_TX_BUFF_SIZE_QUE1				    	TCPIP_GMAC_TX_BUFF_SIZE_DUMMY
#define TCPIP_GMAC_RX_BUFF_COUNT_QUE1				   		1
#define TCPIP_GMAC_RX_BUFF_COUNT_THRESHOLD_QUE1				0
#define TCPIP_GMAC_RX_BUFF_ALLOC_COUNT_QUE1					0

/*** QUEUE 2 Disabled; Dummy Configuration ***/
#define TCPIP_GMAC_TX_DESCRIPTORS_COUNT_QUE2				TCPIP_GMAC_TX_DESCRIPTORS_COUNT_DUMMY
#define TCPIP_GMAC_RX_DESCRIPTORS_COUNT_QUE2				TCPIP_GMAC_RX_DESCRIPTORS_COUNT_DUMMY
#define TCPIP_GMAC_RX_BUFF_SIZE_QUE2				    	TCPIP_GMAC_RX_BUFF_SIZE_DUMMY
#define TCPIP_GMAC_TX_BUFF_SIZE_QUE2				    	TCPIP_GMAC_TX_BUFF_SIZE_DUMMY
#define TCPIP_GMAC_RX_BUFF_COUNT_QUE2				   		1
#define TCPIP_GMAC_RX_BUFF_COUNT_THRESHOLD_QUE2				0
#define TCPIP_GMAC_RX_BUFF_ALLOC_COUNT_QUE2					0

/*** QUEUE 3 Disabled; Dummy Configuration ***/
#define TCPIP_GMAC_TX_DESCRIPTORS_COUNT_QUE3				TCPIP_GMAC_TX_DESCRIPTORS_COUNT_DUMMY
#define TCPIP_GMAC_RX_DESCRIPTORS_COUNT_QUE3				TCPIP_GMAC_RX_DESCRIPTORS_COUNT_DUMMY
#define TCPIP_GMAC_RX_BUFF_SIZE_QUE3				    	TCPIP_GMAC_RX_BUFF_SIZE_DUMMY
#define TCPIP_GMAC_TX_BUFF_SIZE_QUE3				    	TCPIP_GMAC_TX_BUFF_SIZE_DUMMY
#define TCPIP_GMAC_RX_BUFF_COUNT_QUE3				   		1
#define TCPIP_GMAC_RX_BUFF_COUNT_THRESHOLD_QUE3				0
#define TCPIP_GMAC_RX_BUFF_ALLOC_COUNT_QUE3					0

/*** QUEUE 4 Disabled; Dummy Configuration ***/
#define TCPIP_GMAC_TX_DESCRIPTORS_COUNT_QUE4				TCPIP_GMAC_TX_DESCRIPTORS_COUNT_DUMMY
#define TCPIP_GMAC_RX_DESCRIPTORS_COUNT_QUE4				TCPIP_GMAC_RX_DESCRIPTORS_COUNT_DUMMY
#define TCPIP_GMAC_RX_BUFF_SIZE_QUE4				    	TCPIP_GMAC_RX_BUFF_SIZE_DUMMY
#define TCPIP_GMAC_TX_BUFF_SIZE_QUE4				    	TCPIP_GMAC_TX_BUFF_SIZE_DUMMY
#define TCPIP_GMAC_RX_BUFF_COUNT_QUE4				   		1
#define TCPIP_GMAC_RX_BUFF_COUNT_THRESHOLD_QUE4				0
#define TCPIP_GMAC_RX_BUFF_ALLOC_COUNT_QUE4					0

/*** QUEUE 5 Disabled; Dummy Configuration ***/
#define TCPIP_GMAC_TX_DESCRIPTORS_COUNT_QUE5				TCPIP_GMAC_TX_DESCRIPTORS_COUNT_DUMMY
#define TCPIP_GMAC_RX_DESCRIPTORS_COUNT_QUE5				TCPIP_GMAC_RX_DESCRIPTORS_COUNT_DUMMY
#define TCPIP_GMAC_RX_BUFF_SIZE_QUE5				    	TCPIP_GMAC_RX_BUFF_SIZE_DUMMY
#define TCPIP_GMAC_TX_BUFF_SIZE_QUE5				    	TCPIP_GMAC_TX_BUFF_SIZE_DUMMY
#define TCPIP_GMAC_RX_BUFF_COUNT_QUE5				   		1
#define TCPIP_GMAC_RX_BUFF_COUNT_THRESHOLD_QUE5				0
#define TCPIP_GMAC_RX_BUFF_ALLOC_COUNT_QUE5					0

#define TCPIP_GMAC_RX_MAX_FRAME		    			1536
#define TCPIP_GMAC_RX_FILTERS                       \
                                                    TCPIP_MAC_RX_FILTER_TYPE_BCAST_ACCEPT |\
                                                    TCPIP_MAC_RX_FILTER_TYPE_MCAST_ACCEPT |\
                                                    TCPIP_MAC_RX_FILTER_TYPE_UCAST_ACCEPT |\
                                                    TCPIP_MAC_RX_FILTER_TYPE_CRC_ERROR_REJECT |\
                                                    0
#define TCPIP_GMAC_ETH_OPEN_FLAGS       			\
                                                    TCPIP_ETH_OPEN_AUTO |\
                                                    TCPIP_ETH_OPEN_FDUPLEX |\
                                                    TCPIP_ETH_OPEN_HDUPLEX |\
                                                    TCPIP_ETH_OPEN_100 |\
                                                    TCPIP_ETH_OPEN_10 |\
                                                    TCPIP_ETH_OPEN_MDIX_AUTO |\
                                                    TCPIP_ETH_OPEN_RMII |\
                                                    0

#define TCPIP_INTMAC_MODULE_ID		    			GMAC_BASE_ADDRESS
#define TCPIP_INTMAC_PERIPHERAL_CLK  				150000000

#define DRV_GMAC_RX_CHKSM_OFFLOAD				(TCPIP_MAC_CHECKSUM_NONE)			
#define DRV_GMAC_TX_CHKSM_OFFLOAD				(TCPIP_MAC_CHECKSUM_NONE)		

#define DRV_GMAC_INSTANCES_NUMBER				1
#define DRV_GMAC_NUMBER_OF_QUEUES				6
#define DRV_GMAC_CLIENTS_NUMBER					1
#define DRV_GMAC_INDEX	    	    				1
#define DRV_GMAC_PERIPHERAL_ID					1
#define DRV_GMAC_INTERRUPT_SOURCE				GMAC_IRQn

#define DRV_GMAC_INTERRUPT_MODE        				true
#define DRV_GMAC_RMII_MODE					0



/*** announce Configuration ***/
#define TCPIP_STACK_USE_ANNOUNCE
#define TCPIP_ANNOUNCE_MAX_PAYLOAD 	512
#define TCPIP_ANNOUNCE_TASK_RATE    333
#define TCPIP_ANNOUNCE_NETWORK_DIRECTED_BCAST             			false



/*** UDP Configuration ***/
#define TCPIP_UDP_MAX_SOCKETS		                	10
#define TCPIP_UDP_SOCKET_DEFAULT_TX_SIZE		    	512
#define TCPIP_UDP_SOCKET_DEFAULT_TX_QUEUE_LIMIT    	 	3
#define TCPIP_UDP_SOCKET_DEFAULT_RX_QUEUE_LIMIT			3
#define TCPIP_UDP_USE_POOL_BUFFERS   false
#define TCPIP_UDP_USE_TX_CHECKSUM             			true
#define TCPIP_UDP_USE_RX_CHECKSUM             			true
#define TCPIP_UDP_COMMANDS   false
#define TCPIP_UDP_EXTERN_PACKET_PROCESS   false


                                                  
#define TCPIP_INTMAC_PHY_CONFIG_FLAGS              	( 0 \
                                                    | DRV_ETHPHY_CFG_RMII \
                                                    )

#define TCPIP_INTMAC_PHY_LINK_INIT_DELAY  			500
#define TCPIP_INTMAC_PHY_ADDRESS		    			0
#define DRV_ETHPHY_INSTANCES_NUMBER					1
#define DRV_ETHPHY_CLIENTS_NUMBER					1
#define DRV_ETHPHY_INDEX		        			1
#define DRV_ETHPHY_PERIPHERAL_ID					1
#define DRV_ETHPHY_NEG_INIT_TMO		    			1
#define DRV_ETHPHY_NEG_DONE_TMO		    			2000
#define DRV_ETHPHY_RESET_CLR_TMO					500




// *****************************************************************************
// *****************************************************************************
// Section: Application Configuration
// *****************************************************************************
// *****************************************************************************


//DOM-IGNORE-BEGIN
#ifdef __cplusplus
}
#endif
//DOM-IGNORE-END

#endif // CONFIGURATION_H
/*******************************************************************************
 End of File
*/
